Build failure for Storybook with Typescript configuration in freshly generated project

### Current Behavior

In a fresh integrated workspace, when running the generators to create a fresh React library or web app project and set up Storybook, subsequent attempts to build Storybook fail if Typescript was chosen as the configuration file format for Storybook. (That is, the auto-generated `build-storybook` target fails, and in addition, the `storybook` target fails with the same error message.)

The error messages imply that the `@nrwl/storybook:build` executor does not compile the auto-generated *.ts configuration files and instead treats them as Javascript. It does not matter whether the recommended `@nrwl/react:storybook-configuration` generator or the `@nrwl/storybook:configuration` generator is used to set up Storybook for the project.

With the `@nrwl/react:storybook-configuration` generator:

```bash
npx create-nx-workspace@latest nx-integrated-ts-sb --preset=ts --nxCloud=false
cd nx-integrated-ts-sb
npm install -D @nrwl/react
npx nx g @nrwl/react:lib react-library --bundler=vite --style=scss --unitTestRunner=none
npx nx g @nrwl/react:storybook-configuration react-library --tsConfiguration=true --configureCypress=false --generateStories=true --generateCypressSpecs=false
npx nx build-storybook react-library
# Error:
# nx-integrated-ts-sb/packages/react-library/.storybook/main.ts: Unexpected token, expected "from" (3:12)
```

With the `@nrwl/storybook:configuration` generator:

```bash
npx create-nx-workspace@latest nx-integrated-ts-sb --preset=ts --nxCloud=false
cd nx-integrated-ts-sb
npm install -D @nrwl/react @nrwl/storybook
npx nx g @nrwl/react:lib react-library --bundler=vite --style=scss --unitTestRunner=none
npx nx g @nrwl/storybook:configuration react-library --uiFramework=@storybook/react --tsConfiguration=true --configureCypress=false --bundler=vite
npx nx build-storybook react-library
# Error:
# nx-integrated-ts-sb/packages/react-library/.storybook/main.ts: Unexpected token, expected "from" (3:12)
```


### Expected Behavior

The `build-storybook` and `storybook` targets generated by the `@nrwl/react:storybook-configuration` and `@nrwl/storybook:configuration` are expected to work.

If the `--tsConfiguration=true` flag is changed to `--tsConfiguration=false` in the above examples, the Storybook targets work as expected.

### GitHub Repo

https://github.com/OttoG/nx-storybook-ts-error

### Steps to Reproduce

1. `git clone https://github.com/OttoG/nx-storybook-ts-error`
2. `cd nx-storybook-ts-error/nx-integrated-ts-sb`
3. `npm install`
4. `npx nx build-storybook react-library`


### Nx Report

```shell
>  NX   Report complete - copy this into the issue template

   Node : 18.14.0
   OS   : darwin x64
   npm  : 9.3.1
   
   nx : 15.6.3
   @nrwl/angular : Not Found
   @nrwl/cypress : 15.6.3
   @nrwl/detox : Not Found
   @nrwl/devkit : 15.6.3
   @nrwl/esbuild : Not Found
   @nrwl/eslint-plugin-nx : 15.6.3
   @nrwl/expo : Not Found
   @nrwl/express : Not Found
   @nrwl/jest : Not Found
   @nrwl/js : 15.6.3
   @nrwl/linter : 15.6.3
   @nrwl/nest : Not Found
   @nrwl/next : Not Found
   @nrwl/node : Not Found
   @nrwl/nx-cloud : Not Found
   @nrwl/nx-plugin : Not Found
   @nrwl/react : 15.6.3
   @nrwl/react-native : Not Found
   @nrwl/rollup : Not Found
   @nrwl/schematics : Not Found
   @nrwl/storybook : 15.6.3
   @nrwl/web : Not Found
   @nrwl/webpack : Not Found
   @nrwl/workspace : 15.6.3
   @nrwl/vite : 15.6.3
   typescript : 4.8.4
   ---------------------------------------
   Local workspace plugins:
   ---------------------------------------
   Community plugins:
```


### Failure Logs

```shell
> nx run react-library:build-storybook

>  NX  ui framework: @storybook/react

>  NX  Storybook builder starting ...

info => Cleaning outputDir: ./nx-storybook-ts-error/nx-integrated-ts-sb/dist/storybook/react-library

 >  NX   ./nx-storybook-ts-error/nx-integrated-ts-sb/packages/react-library/.storybook/main.ts: Unexpected token, expected "from" (3:12)

   
     1 | import { rootMain } from '../../../.storybook/main';
     2 |
   > 3 | import type { StorybookConfig, Options } from '@storybook/core-common';
       |             ^
     4 |
     5 | import { mergeConfig } from 'vite';
     6 | import viteTsConfigPaths from 'vite-tsconfig-paths';
   Pass --verbose to see the stacktrace.


————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

 >  NX   Ran target build-storybook for project react-library (2s)
 
    ✖    1/1 failed
    ✔    0/1 succeeded [0 read from cache]
```


### Additional Information

_No response_

Hi @mandarini thank you for the very speedy replies!

However, first of all, I am indeed able to build and run Storybook with Node 18, as long as the Storybook configuration is JS-based. (That is, with `--tsConfiguration=false` when configuring Storybook.) I just have to run this before building:

```bash
export NODE_OPTIONS=--openssl-legacy-provider
```

At least in simple cases such as building and running Storybook for the single component that is added by the generators, it works fine with Node 18. I was going to mention that in the original report but forgot about it – sorry!

However, after installing `ts-node`, I get a different kind of error:

```
 >  NX   ⨯ Unable to compile TypeScript:

   packages/react-library/.storybook/main.ts:17:9 - error TS2322: Type '{ core: { builder: string; disableWebpackDefaults?: boolean | undefined; channelOptions?: Partial<Options> | undefined; disableProjectJson?: boolean | undefined; disableTelemetry?: boolean | undefined; enableCrashReports?: boolean | undefined; crossOriginIsolated?: boolean | undefined; }; ... 11 more ...; previewAnn...' is not assignable to type 'StorybookConfig'.
     Object literal may only specify known properties, and 'viteFinal' does not exist in type 'StorybookConfig'.
   
   17   async viteFinal(config: any) {
              ~~~~~~~~~
   
   Pass --verbose to see the stacktrace.


 ————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

 >  NX   Ran target build-storybook for project react-library (3s)
 
    ✖    1/1 failed
    ✔    0/1 succeeded [0 read from cache]
```

VS Code complains about the same issue in the editor. It has done so all the time in my environment, but I was not sure if that would prevent the build from working:

![image](https://user-images.githubusercontent.com/5548343/217052886-bd2e9a76-1b20-433b-a29c-d7430e964388.png)

--- a/packages/storybook/src/executors/build-storybook/build-storybook.impl.ts
+++ b/packages/storybook/src/executors/build-storybook/build-storybook.impl.ts
@@ -56,6 +56,13 @@ function runInstance(options: CLIOptions, storybook7: boolean): Promise<void> {
       mode: 'static',
     } as any); // TODO(katerina): Change to actual types when Storybook 7
   } else {
+    const nodeVersion = process.version.slice(1).split('.');
+    if (+nodeVersion[0] === 18) {
+      logger.warn(`
+        If you are using the @storybook/builder-vite you may experience issues with Node 18.
+        Please use Node 16 if you are using @storybook/builder-vite. 
+      `);
+    }
     return build.buildStaticStandalone({
       ...options,
       ci: true,

--- a/packages/storybook/src/generators/configuration/configuration.ts
+++ b/packages/storybook/src/generators/configuration/configuration.ts
@@ -37,6 +37,7 @@ import {
   storybookSwcAddonVersion,
   storybookTestRunnerVersion,
   storybookVersion,
+  tsNodeVersion,
 } from '../../utils/versions';
 
 export async function configurationGenerator(
@@ -219,6 +220,7 @@ export async function configurationGenerator(
         {},
         {
           ['@storybook/core-common']: storybookVersion,
+          ['ts-node']: tsNodeVersion,
         }
       )
     );

--- a/packages/storybook/src/generators/configuration/project-files-ts/.storybook/main.ts__tmpl__
+++ b/packages/storybook/src/generators/configuration/project-files-ts/.storybook/main.ts__tmpl__
@@ -34,17 +34,19 @@ const config: StorybookConfig = {
         }),
       ],
     });
-  },<% } %><% if (!usesVite) { %>,
-  webpackFinal: async (config, { configType }: Options) => {
-    // apply any global webpack configs that might have been specified in .storybook/main.ts
-    if (rootMain.webpackFinal) {
-      config = await rootMain.webpackFinal(config, { configType } as Options);
-    }
-
-    // add your own webpack tweaks if needed
-
-    return config;
   },<% } %>
-};
+} as StorybookConfig;
 
 module.exports = config;
+
+<% if(!usesVite) { %>
+// To customize your webpack configuration you can use the webpackFinal field.
+// Check https://storybook.js.org/docs/react/builders/webpack#extending-storybooks-webpack-config
+// and https://nx.dev/packages/storybook/documents/custom-builder-configs
+<% } %>
+
+<% if(usesVite) { %>
+// To customize your Vite configuration you can use the viteFinal field. 
+// Check https://storybook.js.org/docs/react/builders/vite#configuration
+// and https://nx.dev/packages/storybook/documents/custom-builder-configs
+<% } %>

--- a/packages/storybook/src/generators/configuration/project-files/.storybook/main.js__tmpl__
+++ b/packages/storybook/src/generators/configuration/project-files/.storybook/main.js__tmpl__
@@ -34,21 +34,17 @@ module.exports = {
         }),
       ],
     });
-  },<% } %><% if (!usesVite) { %>,
-  webpackFinal: async (config, { configType }) => {
-    // apply any global webpack configs that might have been specified in .storybook/main.js
-    if (rootMain.webpackFinal) {
-      config = await rootMain.webpackFinal(config, { configType });
-    }
-    
-    <% if (existsRootWebpackConfig) { %>// for backwards compatibility call the `rootWebpackConfig`
-      // this can be removed once that one is migrated fully to
-      // use the `webpackFinal` property in the `main.js` file
-      config = rootWebpackConfig({ config });
-    <% } %>
-
-    // add your own webpack tweaks if needed
-
-    return config;
   },<% } %>
 };
+
+<% if(!usesVite) { %>
+// To customize your webpack configuration you can use the webpackFinal field.
+// Check https://storybook.js.org/docs/react/builders/webpack#extending-storybooks-webpack-config
+// and https://nx.dev/packages/storybook/documents/custom-builder-configs
+<% } %>
+
+<% if(usesVite) { %>
+// To customize your Vite configuration you can use the viteFinal field. 
+// Check https://storybook.js.org/docs/react/builders/vite#configuration
+// and https://nx.dev/packages/storybook/documents/custom-builder-configs
+<% } %>

--- a/packages/storybook/src/utils/versions.ts
+++ b/packages/storybook/src/utils/versions.ts
@@ -3,8 +3,8 @@ export const storybookVersion = '^6.5.15';
 export const babelCoreVersion = '7.12.13';
 export const babelLoaderVersion = '8.1.0';
 export const babelPresetTypescriptVersion = '7.12.13';
-export const svgrVersion = '^5.4.0';
-export const urlLoaderVersion = '^3.0.0';
+export const svgrVersion = '^6.1.2';
+export const urlLoaderVersion = '^4.1.1';
 export const webpack5Version = '^5.64.0';
 export const viteBuilderVersion = '^0.2.6';
 export const storybookReactNativeVersion = '^6.0.1-beta.11';
@@ -14,5 +14,6 @@ export const storybookNextAddonVersion = '^1.6.6';
 export const storybookTestRunnerVersion = '^0.7.2';
 export const litHtmlVersion = '^2.3.1';
 export const htmlWebpackPluginVersion = '^5.5.0';
+export const tsNodeVersion = '10.9.1';
 
 export const storybook7Version = '^7.0.0-beta.29';

